go/project_structure/library.md

```diff
@@ -0,0 +1,151 @@
+# Library
+
+This page describes common code structure for our libraries.
+
+The example below is taken from [k8sclient] on 2020/04/16. Most of the files
+and directories are elided for simplification.
+
+```
+github.com/giantswarm/k8sclient/
+├── CHANGELOG.md
+├── README.md
+├── go.mod
+├── go.sum
+└── pkg
+    ├── k8sclient
+    │   ├── clients.go
+    │   ├── error.go
+    │   ├── setup.go
+    │   ├── spec.go
+    │   ├── spec_test.go
+    │   └── types.go
+    ├── k8sclienttest
+    │   └── clients.go
+    ├── k8scrdclient
+    │   ├── crd_client.go
+    │   ├── crd_client_test.go
+    │   ├── error.go
+    │   └── spec.go
+    ├── k8srestconfig
+    │   ├── error.go
+    │   └── rest_config.go
+    └── k8sversion
+        ├── error.go
+        ├── funcs.go
+        ├── funcs_test.go
+        └── types.go
+```
+
+Below the go.mod file and a few packages discussed.
+
+## go.mod file
+
+Since we use Go modules releasing a major version of the library means we also
+need to change the import path of the module. The last segment of the module
+import must be `/vX` for `vX.Y.Z` releases where `X >= 2`. E.g. for release
+`v3.4.5` it must be `v3`. This is important for the libraries as they are
+imported in external repositories (as opposed to operators for example) and
+they wouldn't be Go module compatible otherwise. It's worth mentioning that
+module defined with `/vX` segment is treated as a separate module from the v0
+and v1 versions and they can be imported at the same time.
+
+Let's try to illustrate that with `github.com/giantswarm/k8sclient` library
+example.
+
+For releases `v0.x.x` and `v1.x.x` in the `go.mod` file there should be a line
+like:
+
+```
+module github.com/giantswarm/k8sclient
+```
+
+For releases `v2.x.x`:
+
+```
+module github.com/giantswarm/k8sclient/v2
```

Comment:
What happens if this is still `module github.com/giantswarm/k8sclient` and we tag the library as `v2.0.0`? It will just fail to import?

Reply:
It will fail to import talking about incompatibility.